docs/design/features/byreflike-generics.md

```diff
@@ -110,23 +113,123 @@ throw
 
 Adding `gpAcceptByRefLike` to the metadata of a Generic parameter will be considered a non-breaking binary change.
 
-Enumerating of constructors/methods on `Span<T>` and `ReadOnlySpan<T>` may throw `TypeLoadException` if `T` is a ByRefLike type. See "Troublesome APIs" above for the list of APIs that cause this condition.
+Enumerating of constructors/methods on `Span<T>` and `ReadOnlySpan<T>` may throw `TypeLoadException` if `T` is a ByRefLike type. See "Troublesome API mitigation" above for the list of APIs that cause this condition.
+
+## <a name="invalid_il_options"></a> Options for invalid IL
+
+There are two potential options below for how to address this issue.
+
+The first indented IL sequences below represents the `is-type` sequence. Combining the first with the second indented section represents the "type pattern matching" scenario in C#. The below sequence performs a type check and then, if successful, consumes the unboxed instance.
+
+```IL
+// Type check
+ldarg.0
+    box <Source>
+    isinst <Target>
+    brfalse.s NOT_INST
+
+// Unbox and store unboxed instance
+ldarg.0
+    box <Source>
+    isinst <Target>
+    unbox.any <Target>
+stloc.X
+
+NOT_INST:
+ret
+```
+
+With the above IL composition implemented, the following C# describes the following "type pattern matching" scenarios and what one might expect given current C# semantics.
+
+```csharp
+struct S {}
+struct S<T> {}
+ref struct RS {}
+ref struct RS<T> {}
+interface I {}
+class C {}
+class C<T> {}
+
+// Not currently valid C#
+void M<T, U>(T t) where T: allows ref struct
+{
+    // Valid
+    if (t is int i)
+
+    if (t is S s)
+    if (t is S<char> sc)
+    if (t is S<U> su)
+
+    if (t is RS rs)
+    if (t is RS<char> rsc)
+    if (t is RS<U> rsu)
+
+    if (t is string str)
+    if (t is C c)
+    if (t is C<I> ci)
+    if (t is C<U> cu)
+
+    // Can be made to work in IL.
+    if (t is I itf) // A new local "I" would not be used for ByRefLike scenarios.
+                    // The local would be the ByRefLike type, not "I".
+
+    // Invalid
+    if (t is object o)  // ByRefLike types evaluate "true" for object.
+    if (t is U u)
+}
+```
+
+### Option 1) Compiler helpers
+
+The following two helper functions could be introduced and would replace currently invalid `is-type` IL sequences when ByRefLike types are involved. Their behavior would broadly be defined to operate as if the ByRefLike aspect of either the `TFrom` and `TTo` is not present. An alternative approach would be consult with the Roslyn team and define the semantics of these functions to adhere to C# language rules.
```

Comment:
Should we agree on that we are going to do the APIs and that the API behavior will match the needs of the C# language/compiler? Or is it too early for that?

Reply:
I prefer the API approach. I think @jaredpar and @AlekseyTs also are on board with the API approach. Looking for confirmation here.

Reply:
I prefer well defined runtime APIs that act as defacto IL instructions vs taking a sequence of existing IL instructions and assigning a behavior to them.

Reply:
Done. Added note for .NET 10, we will adopt the API approach. Will submit new APIs and tag Roslyn team.